push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/cache/SimpleApnsClientCache.java

```diff
@@ -70,12 +70,29 @@ public void operationComplete(Future<? super Void> future) throws Exception {
                 }).build();
     }
 
-    public ApnsClient getApnsClientForVariantID(final String variantID) {
-        return apnsClientExpiringMap.get(variantID);
+    public ApnsClient getApnsClientForVariantID(final String variantID, final ServiceConstructor<ApnsClient> constructor) {
+        ApnsClient client = apnsClientExpiringMap.get(variantID);
+
+        if (client == null) {
+            logger.debug(String.format("no cached connection for %s, establishing it", variantID));
+            synchronized (apnsClientExpiringMap) {
+                client = constructor.construct();
+                putApnsClientForVariantID(variantID, client);
+
+                return client; // return the newly connected client
+            }
+        } else {
+            logger.debug(String.format("reusing cached connection for %s", variantID));
+            return client; // we had it already
+        }
     }
 
-    public void putApnsClientForVariantID(final String variantID, final ApnsClient apnsClient) {
-        apnsClientExpiringMap.put(variantID, apnsClient);
+    private void putApnsClientForVariantID(final String variantID, final ApnsClient apnsClient) {
+        final ApnsClient client = apnsClientExpiringMap.putIfAbsent(variantID, apnsClient);
+        if (client != null) {
+            logger.warn("duplicate connection in pool, shutting it down");
+            tearDownApnsHttp2Connection(client);
```

Comment:
In my understanding the new value is inserted only if the key is not already associated with another value.
So, the returned value is the old one and, in the map, the old one is present.

> If the specified key is not already associated with a value, associate it with the given value.